Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp

```diff
@@ -0,0 +1,351 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "BackgroundFetch.h"
+
+#if ENABLE(SERVICE_WORKER)
+
+#include "BackgroundFetchInformation.h"
+#include "BackgroundFetchRecordInformation.h"
+#include "CacheQueryOptions.h"
+#include "ExceptionData.h"
+#include "SWServerRegistration.h"
+
+namespace WebCore {
+
+BackgroundFetch::BackgroundFetch(SWServerRegistration& registration, const String& identifier, Vector<BackgroundFetchRequest>&& requests, BackgroundFetchOptions&& options, Ref<BackgroundFetchStore>&& store, NotificationCallback&& notificationCallback)
+    : m_identifier(identifier)
+    , m_options(WTFMove(options))
+    , m_registrationKey(registration.key())
+    , m_registrationIdentifier(registration.identifier())
+    , m_store(WTFMove(store))
+    , m_notificationCallback(WTFMove(notificationCallback))
+    , m_origin { m_registrationKey.topOrigin(), SecurityOriginData::fromURL(m_registrationKey.scope()) }
+{
+    size_t index = 0;
+    m_records.reserveInitialCapacity(requests.size());
+    for (auto& request : requests)
+        m_records.uncheckedAppend(Record::create(*this, WTFMove(request), index++));
+    this->store();
+}
+
+BackgroundFetch::~BackgroundFetch()
+{
+    abort();
+}
+
+BackgroundFetchInformation BackgroundFetch::information() const
+{
+    return { m_registrationIdentifier, m_identifier, m_uploadTotal, m_currentUploadSize, downloadTotal(), m_currentDownloadSize, m_result, m_failureReason, m_recordsAvailableFlag };
+}
+
+void BackgroundFetch::match(const RetrieveRecordsOptions& options, MatchBackgroundFetchCallback&& callback)
+{
+    WebCore::CacheQueryOptions queryOptions { options.ignoreSearch, options.ignoreMethod, options.ignoreVary };
+
+    Vector<Ref<Record>> records;
+    for (auto& record : m_records) {
+        if (options.request.url().isNull() || record->isMatching(options.request, queryOptions))
+            records.append(record);
+    }
+
+    callback(WTFMove(records));
+}
+
+bool BackgroundFetch::abort()
+{
+    if (m_abortFlag)
+        return false;
+
+    m_abortFlag = true;
+    m_isActive = false;
+
+    for (auto& record : m_records)
+        record->abort();
+
+    updateBackgroundFetchStatus(BackgroundFetchResult::Failure, BackgroundFetchFailureReason::Aborted);
+    return true;
+}
+
+void BackgroundFetch::perform(const CreateLoaderCallback& createLoaderCallback)
+{
+    m_currentDownloadSize = 0;
+    for (auto& record : m_records)
```

Comment:
ditto.

Reply:
This is currently guaranteed since m_records is never appending nor removing records outside the constructor.